### ISSUE TITLE
metamorphic: TestMeta failure with range keys

### ISSUE THREAD

**Issue by sumeerbhola** [cockroachdb/pebble#2059]
This was produced using https://github.com/sumeerbhola/pebble/tree/seekge_bpf_hacks which has the fix from https://github.com/cockroachdb/pebble/pull/2051 plus a one line change to undo that fix in SeekGE and some hacky changes to TestMeta to trigger more two level indices. I was trying to reproduce that bug, but instead had a test failure that looks like a range key issue (while running on gceworker).

Unfortunately it does not reproduce every time (it happened twice and then stopped happening -- I am running `go test -exec 'stress -p 1' -timeout 0 -test.v -run TestMeta$ -seed 1666763855951614372` to try to get another reproduction. But documenting it here in case someone can spot something obviously wrong (including with the test).

The failure produced the following diff
```
===== SEED =====
1666763855951614372
===== DIFF =====
_meta/221026-055735.951/{standard-000,random-024}
@@ -3632,13 +3632,13 @@
 iter36.Next("") // [true,"swodmt",<no point>,["swodmt","thjelmejtjj")=>{"@6"="vzbkesallmwcgqkcst"}*] <nil> #3631
 iter35.First() // [true,"cdoiwoa@4","onoouxal",<no range>] <nil> #3632
 db.SingleDelete("vkzilus@1", false /* maybeReplaceDelete */) // <nil> #3633
 iter35.Prev("") // [false] <nil> #3634
 iter35.SeekGE("abisrw@14", "becutssnhp@12") //
```
which isn't very helpful, so I used a hacked up pebble command with CompareHistories (since removing the comments is insufficient -- we also need to sort), to get the following
```
$ pebble meta-diff standard-000/history random-024/history
@@ -3632,13 +3632,13 @@
 iter36.Next("") // [true,"swodmt",<no point>,["swodmt","thjelmejtjj")=>{"@6"="vzbkesallmwcgqkcst"}*] <nil> #3631
 iter35.First() // [true,"cdoiwoa@4","onoouxal",<no range>] <nil> #3632
 db.SingleDelete("vkzilus@1", false /* maybeReplaceDelete */) // <nil> #3633
 iter35.Prev("") // [false] <nil> #3634
 iter35.SeekGE("abisrw@14", "becutssnhp@12") // [invalid] <nil> #3635
-iter36.Prev("") // [true,"rckdqihwl",<no point>,["rckdqihwl","swodmt")=>{"@10"="zzl","@6"="vzbkesallmwcgqkcst"}*] <nil> #3636
+iter36.Prev("") // [true,"swodmt",<no point>,["swodmt","thjelmejtjj")=>{"@6"="vzbkesallmwcgqkcst"}] <nil> #3636
 db.Merge("viih@14", "fvzewizr") // <nil> #3637
-iter36.First() // [true,"rckdqihwl",<no point>,["rckdqihwl","swodmt")=>{"@10"="zzl","@6"="vzbkesallmwcgqkcst"}] <nil> #3638
+iter36.First() // [true,"rckdqihwl",<no point>,["rckdqihwl","swodmt")=>{"@10"="zzl","@6"="vzbkesallmwcgqkcst"}*] <nil> #3638
 db.SingleDelete("fvmpb@7", false /* maybeReplaceDelete */) // <nil> #3639
 db.DeleteRange("kunyhdga@3", "rdthhcm@8") // <nil> #3640
 iter35.First() // [true,"cdoiwoa@4","onoouxal",<no range>] <nil> #3641
 iter35.SetOptions("", "ydedjihi@5", 2 /* key types */, 2, 7, "@13" /* masking suffix */) // <nil> #3642
 iter35.Last() // [true,"xfntdzdvl",<no point>,["xfntdzdvl","ydedjihi@5")=>{""="uj","@13"="zusd","@6"="pknbjpzlbdhpakqskp"}*] <nil> #3643
```

The relevant sequence in standard-000 is (with annotations in <>)
```
iter36 = db.NewIter("", "", 2 /* key types */, 1, 8, "" /* masking suffix */) // <nil> #3579
iter36.Next("") // [false] <nil> #3580
iter36.First() // [true,"alnlnypnrgw",<no point>,["alnlnypnrgw","bbbga")=>{""="wgfkcrplwmysbyv"}*] <nil> #3582
iter36.Next("") // [true,"awfcqozvhn@7","bygax",["alnlnypnrgw","bbbga")=>{""="wgfkcrplwmysbyv"}] <nil> #3587
iter36.SetBounds("", "awfcqozvhn@7") // <nil> #3592
iter36.SeekLT("awfcqozvhn@7", "") // [true,"alnlnypnrgw",<no point>,["alnlnypnrgw","awfcqozvhn@7")=>{""="wgfkcrplwmysbyv"}*] <nil> #3593
iter36.First() // [true,"alnlnypnrgw",<no point>,["alnlnypnrgw","awfcqozvhn@7")=>{""="wgfkcrplwmysbyv"}] <nil> #3596
iter36.First() // [true,"alnlnypnrgw",<no point>,["alnlnypnrgw","awfcqozvhn@7")=>{""="wgfkcrplwmysbyv"}] <nil> #3599
iter36.SeekLT("uqtmqladua@8", "lpeenny@2") // [invalid] <nil> #3600
iter36.SetBounds("kviwwdiheks@15", "mimba@4") // <nil> #3603
iter36.SeekGE("wffvqnayycz@6", "") // [false] <nil> #3611
iter36.Last() // [true,"lprdeic",<no point>,["lprdeic","mimba@4")=>{""="jccniwyrdfhnr","@10"="zzl","@1"="wrimvzy"}*] <nil> #3613
iter36.Prev("") // [false] <nil> #3614
iter36.SeekLT("mimba@4", "") // [true,"lprdeic",<no point>,["lprdeic","mimba@4")=>{""="jccniwyrdfhnr","@10"="zzl","@1"="wrimvzy"}*] <nil> #3617
iter36.Next("") // [false] <nil> #3619
<the SeekLT will not use any point iter positioning optimization> 
iter36.SeekLT("iywnbl@15", "") // [false] <nil> #3622
<following 2 SeekGEs can use the trySeekUsingNext opt for point iter>
iter36.SeekGE("lpeenny@2", "") // [true,"lpeenny@2",<no point>,["kviwwdiheks@15","lprdeic")=>{""="jccniwyrdfhnr","@1"="wrimvzy"}*] <nil> #3627
iter36.SeekGE("mimba@4", "") // [false] <nil> #3628
<bounds are monotonic so monotonic bounds opt in point iter>
iter36.SetBounds("olambclr@9", "vkzilus@1") // <nil> #3629
<only range key, so point iter opt is irrelevant>
iter36.SeekGE("olambclr@9", "") // [true,"rckdqihwl",<no point>,["rckdqihwl","swodmt")=>{"@10"="zzl","@6"="vzbkesallmwcgqkcst"}*] <nil> #3630
<still only range key>
iter36.Next("") // [true,"swodmt",<no point>,["swodmt","thjelmejtjj")=>{"@6"="vzbkesallmwcgqkcst"}*] <nil> #3631
<prev shows the previous range key as expected>
iter36.Prev("") // [true,"rckdqihwl",<no point>,["rckdqihwl","swodmt")=>{"@10"="zzl","@6"="vzbkesallmwcgqkcst"}*] <nil> #3636
```

And in random-024
```
iter36 = db.NewIter("", "", 2 /* key types */, 1, 8, "" /* masking suffix */) // <nil> #3579
iter36.Next("") // [false] <nil> #3580
iter36.First() // [true,"alnlnypnrgw",<no point>,["alnlnypnrgw","bbbga")=>{""="wgfkcrplwmysbyv"}*] <nil> #3582
iter36.Next("") // [true,"awfcqozvhn@7","bygax",["alnlnypnrgw","bbbga")=>{""="wgfkcrplwmysbyv"}] <nil> #3587
iter36.SetBounds("", "awfcqozvhn@7") // <nil> #3592
iter36.SeekLT("awfcqozvhn@7", "") // [true,"alnlnypnrgw",<no point>,["alnlnypnrgw","awfcqozvhn@7")=>{""="wgfkcrplwmysbyv"}*] <nil> #3593
iter36.First() // [true,"alnlnypnrgw",<no point>,["alnlnypnrgw","awfcqozvhn@7")=>{""="wgfkcrplwmysbyv"}] <nil> #3596
iter36.First() // [true,"alnlnypnrgw",<no point>,["alnlnypnrgw","awfcqozvhn@7")=>{""="wgfkcrplwmysbyv"}] <nil> #3599
iter36.SeekLT("uqtmqladua@8", "lpeenny@2") // [invalid] <nil> #3600
iter36.SetBounds("kviwwdiheks@15", "mimba@4") // <nil> #3603
iter36.SeekGE("wffvqnayycz@6", "") // [false] <nil> #3611
iter36.Last() // [true,"lprdeic",<no point>,["lprdeic","mimba@4")=>{""="jccniwyrdfhnr","@10"="zzl","@1"="wrimvzy"}*] <nil> #3613
iter36.Prev("") // [false] <nil> #3614
iter36.SeekLT("mimba@4", "") // [true,"lprdeic",<no point>,["lprdeic","mimba@4")=>{""="jccniwyrdfhnr","@10"="zzl","@1"="wrimvzy"}*] <nil> #3617
iter36.Next("") // [false] <nil> #3619
iter36.SeekLT("iywnbl@15", "") // [false] <nil> #3622
iter36.SeekGE("lpeenny@2", "") // [true,"lpeenny@2",<no point>,["kviwwdiheks@15","lprdeic")=>{""="jccniwyrdfhnr","@1"="wrimvzy"}*] <nil> #3627
iter36.SeekGE("mimba@4", "") // [false] <nil> #3628
iter36.SetBounds("olambclr@9", "vkzilus@1") // <nil> #3629
iter36.SeekGE("olambclr@9", "") // [true,"rckdqihwl",<no point>,["rckdqihwl","swodmt")=>{"@10"="zzl","@6"="vzbkesallmwcgqkcst"}*] <nil> #3630
iter36.Next("") // [true,"swodmt",<no point>,["swodmt","thjelmejtjj")=>{"@6"="vzbkesallmwcgqkcst"}*] <nil> #3631
<shows the same range key>
iter36.Prev("") // [true,"swodmt",<no point>,["swodmt","thjelmejtjj")=>{"@6"="vzbkesallmwcgqkcst"}] <nil> #3636
```

**Comment by sumeerbhola** [comment]
stress reproduces this "20m40s: 59 runs so far, 5 failures (8.47%)". But in 120+ runs without undoing the bug fix it has not reproduced. I may be missing something about the interaction between point and range key iters.

**Comment by jbowens** [comment]
Interestingly, setting `-max-threads 1`, this failed to reproduce over 200 runs. Without constraining the number of threads, it reproduces readily.

**Comment by jbowens** [comment]
I'm suspicious that there's some interference going on between `iter35` and `iter36`. I'm at a loss as to what it could be. `-race` doesn't pick up any data races.

**Comment by jbowens** [comment]
A successful run:
```
  [i = 0xc0001ba500] i.Prev called; before executing i.Key() = "swodmt", i.RangeBounds() = ["swodmt", "thjelmejtjj")
  [i = 0xc0001ba500] i.Prev switching directions
  [i = 0xc0001ba500] i.iterKey != nil (i.iterKey = "tddq@1#1100,0"): calling prevUserKey
  [i = 0xc0001ba500] stepAgain=true: (i.iterKey = "swodmt#72057594037927935,21") calling prevUserKey
iter35.SeekLT("ulrkahewp@8", "kyayy") // [invalid] <nil> #3694
  [i = 0xc0001ba500] i.Prev called; before executing i.Key() = "suaqfsd@12", i.RangeBounds() = ["rckdqihwl", "swodmt")
iter36.Prev("") // [true,"rckdqihwl",<no point>,["rckdqihwl","swodmt")=>{"@10"="zzl","@6"="vzbkesallmwcgqkcst"}*] <nil> #3636
```

A failing run:
```
iter36.Next("") // [true,"swodmt",<no point>,["swodmt","thjelmejtjj")=>{"@6"="vzbkesallmwcgqkcst"}*] <nil> #3631
iter35.Next("") // [false] <nil> #3663
  [i = 0xc0001bc500] i.Prev called; before executing i.Key() = "vtsv", i.RangeBounds() = ["vtsv", "wjpckzwx")
  [i = 0xc0001bc500] i.Prev switching directions
  [i = 0xc0001bc500] i.iterKey == nil
db.SingleDelete("vkzilus@1", false /* maybeReplaceDelete */) // <nil> #3633
  [i = 0xc0001bea00] i.Prev called; before executing i.Key() = "swodmt", i.RangeBounds() = ["swodmt", "thjelmejtjj")
  [i = 0xc0001bea00] i.Prev switching directions
  [i = 0xc0001bea00] i.iterKey != nil (i.iterKey = "thjelmejtjj#72057594037927935,21"): calling prevUserKey
iter35.Prev("") // [false] <nil> #3664
  [i = 0xc0001bea00] stepAgain=true: (i.iterKey = "tddq@1#1100,0") calling prevUserKey
iter36.Prev("") // [true,"swodmt",<no point>,["swodmt","thjelmejtjj")=>{"@6"="vzbkesallmwcgqkcst"}] <nil> #3636
iter35.Next("") // [false] <nil> #3671
```

They vary in the initial position of the iterator when beginning the `Prev`. The failing run's internal iterator is positioned at `thjelmejtjj#72057594037927935,21`. The succeeding run's internal iterator is positioned at `tddq@1#1100,0`.

The failing run observes `tddq@1#1100,0`, at least when Prev-ing backwards, so why isn't it positioned at it? Maybe we're missing the key during forward iteration, so then when switching directions we're off-by-one when we back up. I'll try to verify. I'm not sure why `-max-threads` influences anything though. I suppose it could influence the timing of writes, which might make a particular two-level index structure more likely.

**Comment by jbowens** [comment]
Confirmed that the iterator did not observe `tddq@1#1100,0` in the forward direction, but observed it in reverse, causing the number of Prevs necessary to be off-by-one. Closing this one out as also fixed by #2051.
